src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1507,32 +1519,62 @@
                                ;; Internal system types always use Clojure's type mechanism.
                                (type fact)
                                ;; All other types defer to the provided function.
-                               (fact-type-fn fact))))]
+                               (fact-type-fn fact))))
+
+        ;; Wrap the ancestors-fn so that we don't send internal facts such as NegationResult
+        ;; to user-provided productions.  Note that this work is memoized inside fact-type->roots.
+        wrapped-ancestors-fn (fn [fact-type]
```

Comment:
If you name this "wrapped" should the name `fact-grouping-fn` be changed to `wrapped-fact-type-fn` for consistency?